lib/Transforms/InsertGpuAllocs.cpp

```diff
@@ -0,0 +1,363 @@
+      auto access = it.second;
+      auto loc = alloc.getLoc();
+      builder.setInsertionPoint(alloc);
+      auto gpuAlloc = builder.create<mlir::gpu::AllocOp>(
+          loc, alloc.getType(), /*asyncToken*/ nullptr,
+          /*asyncDependencies*/ llvm::None, alloc.dynamicSizes(),
+          alloc.symbolOperands());
+      alloc->replaceAllUsesWith(gpuAlloc);
+      alloc.erase();
+      if (access.hostRead || access.hostWrite)
+        gpuAlloc->setAttr(imex::getAllocSharedAttrName(),
+                          builder.getUnitAttr());
+    }
+
+    // This is the case where the inputs are passed as arguments to the
+    // function. This code will add the IR for memeory allocation on the device
+    // with gpu.alloc and insert a memref.copy from host to device
+    for (auto it : gpuBufferParams) {
+      auto param = block.getArgument(it.first);
+      auto access = it.second;
+      auto loc = param.getLoc();
+      builder.setInsertionPointToStart(&block);
+      auto memrefType = param.getType().cast<mlir::MemRefType>();
+      auto rank = static_cast<unsigned>(memrefType.getRank());
+      filter.clear();
+      dims.clear();
+      for (auto i : llvm::seq(0u, rank)) {
+        if (memrefType.isDynamicDim(i)) {
+          auto op = builder.create<mlir::memref::DimOp>(loc, param, i);
+          dims.push_back(op);
+          filter.insert(op);
+        }
+      }
+      auto allocType = mlir::MemRefType::get(
+          memrefType.getShape(), memrefType.getElementType(),
+          mlir::MemRefLayoutAttrInterface{}, memrefType.getMemorySpace());
+      auto gpuAlloc = builder.create<mlir::gpu::AllocOp>(
+          loc, allocType, /*asyncToken*/ nullptr,
+          /*asyncDependencies*/ llvm::None, dims,
+          /*symbolOperands*/ llvm::None);
+      auto allocResult = gpuAlloc.getResult(0);
+
+      if (access.hostRead || access.hostWrite)
+        gpuAlloc->setAttr(imex::getAllocSharedAttrName(),
+                          builder.getUnitAttr());
+
+      if (access.hostWrite && access.deviceRead) {
+        auto copy =
+            builder.create<mlir::memref::CopyOp>(loc, param, allocResult);
+        filter.insert(copy);
+      }
+
+      if (allocType != memrefType)
+        allocResult =
+            builder.create<mlir::memref::CastOp>(loc, memrefType, allocResult);
+
+      param.replaceAllUsesExcept(allocResult, filter);
+      builder.setInsertionPoint(term);
+      if (access.hostRead && access.deviceWrite)
+        builder.create<mlir::memref::CopyOp>(loc, allocResult, param);
+
+      if (useGpuDealloc)
+        builder.create<mlir::gpu::DeallocOp>(loc, llvm::None, allocResult);
+      else
+        builder.create<mlir::memref::DeallocOp>(loc, allocResult);
+    }
+  }
+
+  Option<bool> useGpuDealloc{
+      *this, "use-gpu-dealloc",
+      llvm::cl::desc("use gpu.dealloc for gpu allocated memrefs, "
+                     "memref.dealloc will be used otherwise"),
```

Comment:
what is the reason to use memref.dealloc for GPU allocated memrefs?

Reply:
mentioned above